ngraph/core/src/op/gather.cpp

```diff
@@ -128,6 +128,185 @@ shared_ptr<Node> op::v1::Gather::clone_with_new_inputs(const OutputVector& new_a
+        std::vector<Dimension> result_dims(out_rank);
+        PartialShape output_pshape(result_dims);
+
+        // implementation of out_shape formula
+        // data.shape[:batch_dims] + data.shape[batch_dims:axis] + indices.shape[batch_dims:] +
+        // data.shape[axis + 1:]
+        int i = 0;
+        for (; i < batch_dims; i++)
+        {
+            NODE_VALIDATION_CHECK(this,
+                                  data_pshape[i].compatible(indices_pshape[i]),
+                                  "Shapes ",
+                                  data_pshape,
+                                  " and ",
+                                  indices_pshape,
+                                  " are not consistent. data and indices must have equal or "
+                                  "intersecting sizes until batch_dims");
+
+            output_pshape[i] = data_pshape[i] & indices_pshape[i];
+        }
+
+        if (is_axis_set())
+        {
+            int64_t axis = get_axis();
+            for (; i < axis; i++)
+            {
+                output_pshape[i] = data_pshape[i];
+            }
+            for (; i < axis + indices_rank.get_length() - batch_dims; i++)
+            {
+                output_pshape[i] = indices_pshape[batch_dims - axis + i];
+            }
+            for (; i < out_rank; i++)
+            {
+                output_pshape[i] = data_pshape[batch_dims + 1 - indices_rank.get_length() + i];
+            }
+        }
+
+        set_output_type(0, data_type, output_pshape);
+    }
+    else if (data_rank.is_static() && indices_rank.is_static() && batch_dims < 0)
+    {
+        // batch_dims < 0 could be only if axis is not set and in in optimistic scenario
+        // batch_dims will be within the intervals [0, data_rank] && [0, indices_rank]
+        int64_t max_rank = data_rank.get_length() + indices_rank.get_length() - 1;
+        int64_t min_rank = max_rank - max(data_rank.get_length(), indices_rank.get_length());
+
+        set_output_type(0, data_type, PartialShape::dynamic(Dimension(min_rank, max_rank)));
+    }
+    else
+    {
+        set_output_type(0, data_type, PartialShape::dynamic());
+    }
+}
+
+int64_t op::v7::Gather::get_axis() const
+{
+    const auto& const_op = get_constant_from_source(input_value(2));
+    int64_t axis = const_op->cast_vector<int64_t>()[0];
+    if (axis < 0)
+    {
+        const auto& data_rank = get_input_partial_shape(0).rank();
+        if (data_rank.is_static())
+        {
+            axis += data_rank.get_length();
+        }
+    }
+    return axis;
+}
+
+int64_t op::v7::Gather::get_batch_dims() const
+{
+    if (m_batch_dims < 0 && is_axis_set())
+        return get_axis() + m_batch_dims;
+    else
+        return m_batch_dims;
+}
+
+bool op::v7::Gather::is_axis_set() const
+{
+    const auto& axes_constant = get_constant_from_source(input_value(2));
+    if (axes_constant && !axes_constant->cast_vector<int64_t>().empty())
+        return true;
+    return false;
```

Comment:
```cpp
    const auto& axes_constant = get_constant_from_source(input_value(2));
    return axes_constant && !axes_constant->cast_vector<int64_t>().empty();
```

Reply:
Actually, I don't see a reason for the check `axes_constant->cast_vector<int64_t>().empty()` because we already checked that the `axes` input is scalar or 1D tensor with 1 element.

Reply:
Agree than old code was ugly. Agree that there is no need to check for emptiness, removed it. Check for only `get_constant_from_source(input_value(2))` is enough.

But I still keep this function as a separate since ` iis_axis_set()` is more meaningful and readable than `get_constant_from_source(input_value(2))`